File: 621-task-scheduler/task-scheduler.py
class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        hp = []
        waiting = deque()
        currTime = 0
        freqs = defaultdict(int)
        for task in tasks:
            freqs[task]+=1
        for k,v in freqs.items():
            heapq.heappush(hp,(-v,k,0))
        while hp or waiting:
            while waiting and waiting[0][2] <= currTime:
                heapq.heappush(hp,waiting.popleft())
            if hp:
                popped = heapq.heappop(hp)
                if popped[0] < -1:
                    waiting.append((popped[0]+1,popped[1],currTime+n+1))
                currTime+=1
            else:
                currTime = waiting[0][2]
        return currTime

File: 621-task-scheduler/task-scheduler.java
class Solution {
    public int leastInterval(char[] tasks, int n) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());
        Queue<Pair> waitingQueue = new LinkedList<>();

        int[] freqs = new int[26];
        for (char task : tasks) {
            freqs[task - 'A']++;
        }

        for (int f : freqs) {
            if (f > 0) pq.add(f);
        }

        int time = 0;

        while (!pq.isEmpty() || !waitingQueue.isEmpty()) {

            while (!waitingQueue.isEmpty() &&
                   waitingQueue.peek().timeAvailableAt == time) {
                pq.add(waitingQueue.poll().freq);
            }

            if (!pq.isEmpty()) {
                int curr = pq.poll();
                if (curr - 1 > 0) {
                    waitingQueue.add(new Pair(curr - 1, time + n + 1));
                }
            }

            time++;
        }
        return time;
    }
}

class Pair {
    int freq;
    int timeAvailableAt;

    Pair(int freq, int timeAvailableAt) {
        this.freq = freq;
        this.timeAvailableAt = timeAvailableAt;
    }
}
